Feature: Set Installation Status to Warning when all expected packages are installed

feng-j678 opened the pull request:
[x] feature 26928297
[x] Currently when security/critical package(s) installation is failed, the installation operation will mark as an error even after retry package is succeeded. 
[x] new code changes for first time when security/critical package(s) installation failed, but after retry if all package(s) are installed as planned, mark installation operation to Warning and add details in error object that "all supposed package(s) are installed"

<img width="794" alt="image" src="https://github.com/user-attachments/assets/c4f563ac-a688-4438-a6dc-b28707d08437" />
![image](https://github.com/user-attachments/assets/d5d60969-96de-4c15-9e70-d30642e61d8b)

High-level PR comments by @rane-rajasi were in this 'comments PR' - leaving link: https://github.com/Azure/LinuxPatchExtension/pull/314  (this will be closed)

feng-j678 committed:
add logic to set installation warning when all pkgs are installed

feng-j678 committed:
add print in statushandler to help with debug

feng-j678 committed:
add print in patchinstaller

feng-j678 committed:
add unit test to test install successful on retry

feng-j678 committed:
remove print, revert linebreak

feng-j678 committed:
correct spelling

feng-j678 committed:
remove comments

kjohn-msft commented:
Note left offline to clean up miscellaneous noisy line changes to make the PR easier to review.

kjohn-msft reviewed (changes requested):
See comment history - please remove misc noisy lines (whitespace changes and non-functional changes) to make this easier to review.

rane-rajasi reviewed (changes requested):
Please add a description of the current scenario and what this change is trying to achieve,

feng-j678 committed:
Merge branch 'master' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning

feng-j678 committed:
Merge branch 'feature/pkg_failed_retry_set_status_warning' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes

feng-j678 committed:
revert format changes

michellemcdaniel commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +828,45 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+        """ Check if all supposed security and critical packages are installed """
+        # Get the list of installed packages
+        installed_packages_list = self.status_handler.get_installation_packages_list()
+        print('what is installed_packages_list', installed_packages_list)
```
Did you mean to keep this print statement in?

michellemcdaniel commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -416,18 +427,29 @@ def log_final_metrics(self, maintenance_window, patch_installation_successful, m
+        self.__log_progress_status(maintenance_window, installed_update_count)
+
+        message = "\n\nAll supposed package(s) are installed."
+        self.status_handler.add_error_to_status(message, Constants.PatchOperationErrorCodes.PACKAGES_RETRY_SUCCEEDED)
```
I don't know much about this codebase, but we're logging the warnings as errors, despite the fact that our checks are finding that the packages are installed. Is that correct? We want to error, despite the fact that these are warnings and it succeeded on retry?

feng-j678 replied:
Yes, this is not uncommon in the LPE repository, where function names can sometimes be misleading. It seems that initially, the function name accurately reflected its intended action, but over time, as more changes were introduced, naming conventions were not consistently maintained. We are aware of this and actively working to improve it.

rane-rajasi replied:
I would argue the naming in this is still correct and was never changed from it's original implementation. It's named add_error_to_status() since it adds a message to Error objects within the status file.

feng-j678 committed:
Merge branch 'master' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes.

feng-j678 committed:
Merge branch 'feature/pkg_failed_retry_set_status_warning' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull remote branch changes

feng-j678 committed:
create fake comment to reset file format

feng-j678 committed:
Merge branch 'tmp/resolve_installer_format' into feature/pkg_failed_retry_set_status_warning
pull patch installer fake changes to restore format

feng-j678 committed:
create conflict

feng-j678 committed:
create conflict

feng-j678 committed:
Merge branch 'tmp/A' into tmp/B
create conflict to reset format.

feng-j678 committed:
remove print

feng-j678 committed:
add fake comment

feng-j678 committed:
reset format

feng-j678 committed:
Merge branch 'tmp/A' into tmp/B

reset file format

feng-j678 committed:
remove add trailing space

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +828,44 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+    # region - Failed packages retry succeeded
+    def __check_if_all_packages_installed(self):
+        #type (none) -> bool
+        """ Check if all supposed security and critical packages are installed """
```
Check if all **requested** security and critical packages are installed

feng-j678 replied:
done

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +828,44 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+        self.composite_logger.log(progress_status)
+
+    # region - Failed packages retry succeeded
+    def __check_if_all_packages_installed(self):
```
rename to __are_all_packages_installed() or __check_all_requested_packages_install_state()

feng-j678 replied:
__check_all_requested_packages_install_state() done

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +828,44 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+
+        # Return false there's no security/critical packages
+        if len(security_critical_packages) == 0:
+            return False
```
This scenario will fail for CentOS and in future for Azure Linux. Why are we checking against Critical/Security vs 
<img width="443" alt="image" src="https://github.com/user-attachments/assets/efa616cc-bd6c-483f-bfa8-1090ddbfb356" />
from PatchInstallationSummary?

kjohn-msft replied:
@feng-j678 - distinguishing between classifications wasn't in the feature req spec right?

feng-j678 replied:
this came up in one of our scrum meetings, I've altered the code logic to check all pkgs are installed or not

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +828,44 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+        # All security/critical packages are installed
+        return True
+
+    def __sent_metadata_health_store(self):
```
__sen**d**_metadata_health_store

feng-j678 replied:
done

rane-rajasi commented on src/core/src/CoreMain.py:
```diff
@@ -101,6 +101,9 @@ def __init__(self, argv):
                     patch_installer.mark_installation_completed()
                     overall_patch_installation_operation_successful = True
+                elif patch_installer.get_enabled_installation_warning_status():
+                    patch_installer.mark_installation_warning_completed()
```
rename to mark_installation_completed_with_warning()

feng-j678 replied:
done

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -52,6 +52,8 @@ def __init__(self, env_layer, execution_config, composite_logger, telemetry_writ
 
         self.stopwatch = Stopwatch(self.env_layer, self.telemetry_writer, self.composite_logger)
 
+        self.__enable_installation_warning_status = False
```
Why do you need this variable? It has dependencies on other code parts spread all across to mark it's correct state.

feng-j678 replied:
this is a flag variable to use during the final operation to determine status change in coremain.py:

![image](https://github.com/user-attachments/assets/457c639c-50e1-4adb-aa7f-88fe47120e76)

kjohn-msft replied:
Please rename this into something else as suggested.

self.__change_patch_installation_status_to_warning_from_failed = false

Or anything that's very specific.

feng-j678 replied:
set_patch_installation_status_to_warning_from_failed

feng-j678 committed:
B
Merge branch 'master' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes

feng-j678 committed:
refactor functions and comments

feng-j678 commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -416,6 +427,17 @@ def log_final_metrics(self, maintenance_window, patch_installation_successful, m
+    def log_final_warning_metric(self, maintenance_window, installed_update_count):
+        """
+        logs the final metrics for warning installation status.
+        """
```
done

feng-j678 commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -416,6 +427,17 @@ def log_final_metrics(self, maintenance_window, patch_installation_successful, m
+
+        self.__log_progress_status(maintenance_window, installed_update_count)
+
+        message = "\n\nAll supposed package(s) are installed."
```
done

feng-j678 commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -683,12 +705,17 @@ def mark_installation_completed(self):
+        """ Marks Installation operation as warning by updating the status of PatchInstallationSummary as warning and patch metadata to be sent to healthstore.
+        This is set outside of start_installation function to a restriction in CRP, where installation substatus should be marked as warning only after the implicit (2nd) assessment operation
+        and all supposed packages are installed as expected
+        """
```
done

feng-j678 committed:
add extra line btw imports and class

feng-j678 committed:
remove extra spaces

feng-j678 committed:
extract log metric into a function

feng-j678 commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -379,7 +386,12 @@ def install_updates(self, maintenance_window, package_manager, simulate=False):
+            self.log_final_warning_metric(maintenance_window, installed_update_count)
+            self.__enable_installation_warning_status = True
+        else:
+            self.log_final_metrics(maintenance_window, patch_installation_successful, maintenance_window_exceeded, installed_update_count)
```
done

kjohn-msft commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +819,52 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+            self.log_final_metrics(maintenance_window, patch_installation_successful, maintenance_window_exceeded, installed_update_count)
+
+    def __check_all_requested_packages_install_state(self):
+        #type (none) -> bool
```
![image](https://github.com/user-attachments/assets/3b22337d-5a3e-4830-86ba-7f9bb2bb5519)

feng-j678 replied:
done

kjohn-msft commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -416,10 +420,19 @@ def log_final_metrics(self, maintenance_window, patch_installation_successful, m
+
+        self.__log_progress_status(maintenance_window, installed_update_count)
+
+        message = "\n\nAll requested package(s) are installed. Any patch errors marked are from previous attempts."
```
_message = "\n\n_

Please don't add leading newlines into statuses that go into API result bodies.

feng-j678 replied:
remove \n\n

kjohn-msft commented on src/core/src/service_interfaces/StatusHandler.py:
```diff
@@ -262,6 +262,11 @@ def get_os_name_and_version(self):
             return "unknownDist_unknownVer"
+
+    def get_installation_packages_list(self):
+        #type (none) -> list
```
# type () -> list
or
# type (none) -> list

but needs space between # and type. Applies everywhere. IDE will warn about this.

kjohn-msft commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +819,52 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+        self.composite_logger.log(progress_status)
+
+    # region - Failed packages retry succeeded
+    def log_final_error_status_msg(self, patch_installation_successful, maintenance_window_exceeded, maintenance_window, installed_update_count):
+        if not patch_installation_successful and not maintenance_window_exceeded and self.__check_all_requested_packages_install_state():
+            self.log_final_warning_metric(maintenance_window, installed_update_count)
+            self.__enable_installation_warning_status = True
+        else:
+            self.log_final_metrics(maintenance_window, patch_installation_successful, maintenance_window_exceeded, installed_update_count)
```
This function is combining unrelated things:
- final metrics and 
- status writes and other things nested into log_final_warning_metric

One - these concepts shouldn't be combined (single responsibility principle in SOLID principles), and two, the name of the function obscures this combination.

feng-j678 replied:
split it into 2 functions, adding __check_installation_status_can_set_to_warning

kjohn-msft commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -683,12 +696,17 @@ def mark_installation_completed(self):
-                wait_after_update=False)
+        self.__send_metadata_health_store()
+
+    def mark_installation_completed_with_warning(self):
+        """ Marks Installation operation as warning by updating the status of PatchInstallationSummary as warning and patch metadata to be sent to healthstore.
+        This is set outside of start_installation function to a restriction in CRP, where installation substatus should be marked as warning only after the implicit (2nd) assessment operation
+        and all supposed packages are installed as expected. """
+
+        self.status_handler.set_installation_substatus_json(status=Constants.STATUS_WARNING)
+
+        # Update patch metadata in status for auto patching request, to be reported to healthStore
+        self.__send_metadata_health_store()
```
Conceptually very good, but violates single responsibility principle.

kjohn-msft reviewed (changes requested):
Please address the immediate comments including single responsibility violations. Need to look at it again today after these changes.

feng-j678 committed:
Bi
Merge branch 'master' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes

rane-rajasi reviewed (changes requested):
Other comments inline.

Overall a comment on the implementation technique:

According to the pseudo code discussed in this feature design, this change was supposed to be:
-> when overall status is set in exception block, check if installation state is marked as failed and **customer's intent is achieved**
    -> if yes, set install state to warning with a message in error block
    -> if no, do nothing

Which in code could just be a 2 function change, with **no changes to any other part of code (also discussed prior during pseudo code discussion)**:
 -> i.e. when overall status is set in exception block: i.e.
 elif patch_installer.check_if_install_state_should_be_set_to_warning():
      &nbsp;&nbsp;&nbsp;&nbsp; patch_installer.mark_installation_completed_with_warning()
      &nbsp;&nbsp;&nbsp;&nbsp; overall_patch_installation_operation_successful = True

here 
1. check_if_install_state_should_be_set_to_warning() 
&nbsp;&nbsp;&nbsp;&nbsp;1. checks if install state is failed and **customer's intent is achieved**. i.e. checks if all the patches a customer wanted to update, based on the selected classification, inclusion and exclusion lists, are updated (this is what the other comment thread in __check_all_requested_packages_install_state() is referring to, on not simply looking at security and critical in classifications)
&nbsp;&nbsp;&nbsp;&nbsp;2. If yes, this function returns true else false
2. Then mark_installation_completed_with_warning() sets install state to warning along with any error message, logs and metadata for healthstore.

Instead your code has:
1. get_enabled_installation_warning_status(), which only checks a variable '__mark_installation_as_warning'
2. This variable '__mark_installation_as_warning' is set in a log function (that includes changing existing code)
This is adding dependencies in different parts of code, which those different parts are not responsible for, for an action to complete. **I still don't see why existing logging mechanism needs a change here.** @feng-j678, @kjohn-msft can either of you please expand on this.

feng-j678 commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -416,10 +420,19 @@ def log_final_metrics(self, maintenance_window, patch_installation_successful, m
+        """ Log the final metrics for warning installation status. """
+
+        self.__log_progress_status(maintenance_window, installed_update_count)
+
```
done

feng-j678 commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -683,12 +696,17 @@ def mark_installation_completed(self):
-                patch_version=self.execution_config.health_store_id,
-                report_to_healthstore=True,
-                wait_after_update=False)
+        self.__send_metadata_health_store()
```
done

feng-j678 committed:
apply code refactor

feng-j678 committed:
alter logic to check all packages are installed instead critical/securit

feng-j678 committed:
refactor extra lines

feng-j678 commented on src/core/src/CoreMain.py:
```diff
@@ -101,6 +101,9 @@ def __init__(self, argv):
                 if patch_assessment_successful and patch_installation_successful:
                     patch_installer.mark_installation_completed()
                     overall_patch_installation_operation_successful = True
+                elif patch_installer.get_enabled_installation_warning_status():
```
update it to set_patch_installation_status_to_warning_from_failed() to be more descriptive

rane-rajasi replied:
@feng-j678,  set_patch_installation_status_to_warning_from_failed() makes it look like an action the function is performing (i.e. setting patch status) vs a check. The function name to indicate what it's doing. It should be should_set_installation_status_to_warning() or check_if_patch_installation_status_to_be_set_to_warning()

rane-rajasi replied:
@feng-j678 this is not addressed yet

feng-j678 committed:
Merge branch 'master' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -379,7 +384,9 @@ def install_updates(self, maintenance_window, package_manager, simulate=False):
         installed_update_count += self.perform_status_reconciliation_conditionally(package_manager, True)
-        self.log_final_metrics(maintenance_window, patch_installation_successful, maintenance_window_exceeded, installed_update_count)
+
+        self.__enable_installation_status_to_warning_flag = self.__check_installation_status_can_set_to_warning(patch_installation_successful, maintenance_window_exceeded)
```
Why is this being set here?

kjohn-msft replied:
@feng-j678 - marking open comment

feng-j678 replied:
logic update: initialize self.__enable_installation_status_to_warning_flag within __check_installation_status_can_set_to_warning, so no need to initialize on line 388

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +813,45 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+    # region - Failed packages retry succeeded
+    def log_final_installation_metric(self, patch_installation_successful, maintenance_window_exceeded, maintenance_window, installed_update_count):
+        """ log final installation operation status. """
+        if self.__enable_installation_status_to_warning_flag:
```
Instead of setting this variable before this function is called each time, why not check if warning should be set in if:
if self.__check_installation_status_can_set_to_warning(patch_installation_successful, maintenance_window_exceeded)

This would also remove the dependency on that variable

feng-j678 replied:
initialize self.__enable_installation_status_to_warning_flag within __check_installation_status_can_set_to_warning, done

rane-rajasi replied:
@feng-j678 this is still open I believe. The suggestion was to **check if warning should be set** in the function so you don't need to use this variable: self.__enable_installation_status_to_warning_flag

**This would also remove the dependency on that variable**

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +813,45 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+
+    def __check_all_requested_packages_install_state(self):
+        # type (none) -> bool
+        """ Check if any package(s) are all installed. """
```
This isn't a check for '**any**' packages are installed but everything the customer has requested is installed.

feng-j678 replied:
update description to "Check if all requested package(s) are installed.", Done

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +813,45 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+        # type (bool, bool) -> bool
+        return not patch_installation_successful and not maintenance_window_exceeded and self.__check_all_requested_packages_install_state()
+
+    def __check_all_requested_packages_install_state(self):
```
This should be in status handler

feng-j678 replied:
move logic to status handler, done

rane-rajasi reviewed (changes requested):
Comments inline

feng-j678 committed:
Merge branch 'master' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes

feng-j678 committed:
refactor __check_installation_status_can_set_to_warning

feng-j678 committed:
add extra lines

feng-j678 committed:
refactor __send_metadata_to_health_store

feng-j678 committed:
Merge branch 'master' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes.

rane-rajasi commented on src/core/src/CoreMain.py:
```diff
@@ -101,6 +101,9 @@ def __init__(self, argv):
                 if patch_assessment_successful and patch_installation_successful:
                     patch_installer.mark_installation_completed()
                     overall_patch_installation_operation_successful = True
+                elif patch_installer.set_patch_installation_status_to_warning_from_failed():
```
A previous comment on this is still open: https://github.com/Azure/LinuxPatchExtension/pull/291#discussion_r2028921715
<img width="486" alt="image" src="https://github.com/user-attachments/assets/5a9dae1f-5adf-4d8e-83c0-067ee3c3d33a" />

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -683,12 +688,17 @@ def mark_installation_completed(self):
+    def mark_installation_completed_with_warning(self):
+        """ Marks Installation operation as warning by updating the status of PatchInstallationSummary as warning and patch metadata to be sent to healthstore.
+        This is set outside of start_installation function to a restriction in CRP, where installation substatus should be marked as warning only after the implicit (2nd) assessment operation
+        and all supposed packages are installed as expected. """
```
minor: ...**and all expected packages are installed**

feng-j678 replied:
done

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +811,35 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+                                                        "Completed processing packages!")
+        self.composite_logger.log(progress_status)
+        
+    def __send_metadata_to_health_store(self):
```
nit: add a one line function description

feng-j678 replied:
done

rane-rajasi committed:
PR feedback explanation

rane-rajasi committed:
Revert "PR feedback explanation"

This reverts commit f9ae07e6bd1df3701e6c0b6f6b0d54ee394b0cc9.

rane-rajasi reviewed (changes requested):
See this PR commit: https://github.com/Azure/LinuxPatchExtension/pull/314/commits/f9ae07e6bd1df3701e6c0b6f6b0d54ee394b0cc9 
 as reference for what is expected out of this feature.

NOTE: The new UTs you've added are failing due to legacy code from your changes that weren't carried over to this one, please look into those.
I believe we can also avoid returning maintenance_window_exceeded from start_installation() but that would require a lot of existing code changes which again shouldn't happen in here.

feng-j678 committed:
refactor are_all_requested_packages_installed

feng-j678 committed:
Merge branch 'feature/pkg_failed_retry_set_status_warning' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes

feng-j678 committed:
refactor code

feng-j678 committed:
update new UTs

feng-j678 committed:
merge branch 'master' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes.

feng-j678 committed:
add pr 314 changes

feng-j678 committed:
remove unused variable

feng-j678 committed:
merge branch 'master' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes.

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -52,6 +53,7 @@ def __init__(self, env_layer, execution_config, composite_logger, telemetry_writ
 
         self.stopwatch = Stopwatch(self.env_layer, self.telemetry_writer, self.composite_logger)
 
+
```
nit: remove extra line break

feng-j678 replied:
done

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +805,30 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+        message = "All requested package(s) are installed. Any patch errors marked are from previous attempts."
+        self.composite_logger.log(message)
+        self.status_handler.add_error_to_status(message=message, error_code=Constants.PatchOperationErrorCodes.PACKAGES_RETRY_SUCCEEDED, current_operation_override_for_error=Constants.INSTALLATION)
+    
```
Why is [current_operation_override_for_error=Constants.INSTALLATION] explicitly passed to status_handler.add_error_to_status()?

feng-j678 replied:
because in coremain.py, we are invoking assessment logic and in assessment first line of code is set operation to assessment. I was trying to avoid making unnecessary code changes to minimize new regressions so i passed in a parm to override the operation then it will update the correct error json. since you raise this up, I've added a minor condition check in patchassessor.py to check for installation operation, so no overriding.

![image](https://github.com/user-attachments/assets/cb50619e-b283-4004-a562-1b9dea001539)
self.status_handler.set_current_operation(Constants.ASSESSMENT)

rane-rajasi replied:
My other comment applies here as well: https://github.com/Azure/LinuxPatchExtension/pull/291/files#r2070283468.

set current_operation in mark_installation_completed_with_warning(). Try to use similar code as reference and always test the code for all scenarios when introducing a change in existing code.

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +805,30 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+        """ Check if patch installation status can be set to warning from failed. """
+        # type (bool, bool) -> bool
+        return not patch_installation_successful and not maintenance_window_exceeded and self.status_handler.are_all_requested_packages_installed()
+    # endregion
```
Is this the end of this file? If so, make sure to leave 2 empty lines at the end

feng-j678 replied:
done

rane-rajasi replied:
**2 empty lines**

feng-j678 replied:
not sure why, but there are 2 empty lines
![image](https://github.com/user-attachments/assets/f50a678f-eee5-4937-8ba0-0a73dd1b7e40)

rane-rajasi reviewed (changes requested):
Minor comments inline

feng-j678 committed:
add condition to prevent operation override

feng-j678 committed:
Merge branch 'master' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes

feng-j678 commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -324,7 +326,7 @@ def install_updates(self, maintenance_window, package_manager, simulate=False):
             package_and_dependency_versions = [version]
-            
+
             self.include_dependencies(package_manager, [package], [version], all_packages, all_package_versions, packages, package_versions, package_and_dependencies, package_and_dependency_versions)
```
done

feng-j678 committed:
undo line breaks

feng-j678 committed:
add extra line

feng-j678 committed:
remove line break

rane-rajasi commented on src/core/src/core_logic/PatchAssessor.py:
```diff
@@ -42,7 +42,9 @@ def __init__(self, env_layer, execution_config, composite_logger, telemetry_writ
-        self.status_handler.set_current_operation(Constants.ASSESSMENT)
+        # Prevent operation override during installation operation
+        if self.status_handler.get_current_operation() != Constants.INSTALLATION:
+            self.status_handler.set_current_operation(Constants.ASSESSMENT)
```
Why is this change necessary? This is a breaking change, current operation needs to be set to whichever sub operation (i.e. assessment, installation, configure patching) is in progress for the code to be able to update the correct substatus

feng-j678 replied:
change is corrected

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -683,12 +677,20 @@ def mark_installation_completed(self):
+        """ Marks Installation operation as warning by updating the status of PatchInstallationSummary as warning and patch metadata to be sent to healthstore.
+        This is set outside of start_installation function to a restriction in CRP, where installation substatus should be marked as warning only after the implicit (2nd) assessment operation
+        and all customer requested packages are installed. """
+        
```
Refer the function mark_installation_completed()

<img width="749" alt="image" src="https://github.com/user-attachments/assets/1a98916a-4e0f-48d2-8b19-57b4adafe390" />

Incase you want to ensure PatchInstallationSummary substatus is updated by this function, you need to set current_operation to Installation here, instead of changing existing code in PatchAssessor.py

feng-j678 replied:
this is done

rane-rajasi commented on src/core/tests/Test_CoreMain.py:
```diff
@@ -1285,6 +1359,26 @@ def __check_telemetry_events(self, runtime):
+
+        # Check installation status is WARNING
+        self.assertTrue(substatus_file_data[1]["name"] == Constants.PATCH_INSTALLATION_SUMMARY)
+        self.assertTrue(substatus_file_data[1]["status"].lower() == Constants.STATUS_WARNING.lower())
```
Add a Installed patches and/or pending patches assertion here if possible. i.e. say, Pending patch count == 0

feng-j678 replied:
done

rane-rajasi commented on src/core/src/core_logic/PatchInstaller.py:
```diff
@@ -801,3 +803,31 @@ def get_max_batch_size(self, maintenance_window, package_manager):
+    def should_patch_installation_status_be_set_to_warning(self, patch_installation_successful, maintenance_window_exceeded):
+        """ Check if patch installation status can be set to warning from failed. """
+        # type (bool, bool) -> bool
+        return not patch_installation_successful and not maintenance_window_exceeded and self.status_handler.are_all_requested_packages_installed()
```
@kjohn-msft I think we also need to check if reboot is pending here. We have an existing function mark_installation_completed() which also sets installation status to warning based on a few conditions, wondering if we should merge this one with that
<img width="749" alt="image" src="https://github.com/user-attachments/assets/837dee22-8e55-4e29-b1a0-4b9883a70cb8" />

rane-rajasi reviewed (changes requested):
5 open comments inline (2 from previous review) and 1 comment open for discussion

feng-j678 committed:
revert assessor and add set_current_operation

feng-j678 committed:
add assertion for pendingpatchcount

feng-j678 committed:
Merge branch 'master' of https://github.com/Azure/LinuxPatchExtension into feature/pkg_failed_retry_set_status_warning
pull master changes

kjohn-msft commented:
High-level PR comments by @rane-rajasi were in this 'comments PR' - leaving link: https://github.com/Azure/LinuxPatchExtension/pull/314  (this will be closed)
